### PR TITLE
feat(log): bump log module from 1.4.0 to 6.3.1

### PR DESCRIPTION
**Pull request by joeyguerra**
# Intent

Update to the latest version

# Approach

Sinon couldn't wrap the `warning` function in the `log` module when I updated `log` to the latest version. So I created an object to decorate the underlying `log` module.

This means that technically the `logger` API has changed.

# Reasoning

I didn't want to change the tests because, to me, they represent the agreed upon public API. So my thought is that if the tests pass, the probability of breaking a dependent is low.

#1492

### CODE REVIEW

**Comment by levenleven** [comment]
👋 if we're already replacing a logger module would be nice to go with something that supports structured logging like https://www.npmjs.com/package/pino

**Comment by joeyguerra** [comment]
So. I’m already questioning myself suggesting replacing logging with this PR. Questions like:

Is now the time to make such a big move? And not at least bump to a minor version release?

is this even a *big* move?

What minimum version of Node.is does this support?

what minimum version should Hubot support?

I’m unfamiliar with pino, but *structured* logging seems like a common requirement “nowadays”.

what are you’re thoughts?

**Comment by xurizaemon** [comment]
Lots of change has come in lately which is good, but too much can be a problem.

PR description currently doesn't say _why_ were making this change, it describes a problem which occurred. I think "Sinon couldn't wrap the warning function in the log module when I updated log to the latest version." means "some test coverage stopped working", which might suggest replacing the logger (but are there other options like fixing tests? it helps to document why not that approach).

Considering a structured logger is _also_ a good idea and that would be a feature change I think, since it would change one of Hubot's interfaces (for the better). Again I'd like to see the _why_ of it in the issue/PR rather than just the change proposed. Tedious, I know :)

Sidebar: If we had semantic-release configured (#965?) it would put onus on us to stop the "chore" commit message here ([docs](https://semantic-release.gitbook.io/semantic-release/), since "chore" says "nothing to see here" and this is a feature version change IMO?

**Comment by joeyguerra** [comment]
Apologies.

Great points. I'll update the title and description to include the *why* and add my thought process.

I've reverted the change back to "just update the log version" and try an approach that will pass the tests so we can consider moving forward with just that change and create a different PR to discuss replacing `log` with `Pino` for the structured logging aspects.

**Comment by levenleven** [comment]
Yeah, all above makes sense. Sorry for bringing the discussion into PR.

One thing to consider is that newer version of the same `log` package can also have breaking changes and can break stuff on users' side.

> what minimum node version should Hubot support?

I think we should align with https://endoflife.date/nodejs, so currently it would be 16+ (of course we should bump major when we change this requirement)

**Comment by xurizaemon** [comment]
All good @levenleven - thanks for proposing!

@joeyguerra yep if it's just upgrading existing logging component then chore is probably right; if we swapped loggers and Hubot's log output changed, that might impact folks who are parsing it, and therefore I'd argue it's worth some more significant signifier. Which is what I was trying to say - that once those commits have "real world impact" (ie firing `semantic-release`), it puts the requirement on us to consider and describe what we changed, which is hopefully for the better.

All very theoretical until we wire that up again 😁

**Comment by joeyguerra** [comment]
> API is one primary interface of Hubot, but log output is also (IMO) an interface (albeit secondary), and chat responses are another primary interface, etc.

Based on this thought, "log output is also an interface", I'm going to change the PR title to `feat(log)` because I can't say for sure that the log output has NOT changed. I also feel like since we're going from a `1.x` to a `3.x` version of log, that the probability the output has changed is high.

Therefore, I'd like to get the other PRs merged first so that we can do another release. Then consider merging this one after that for a more impactful release.

**Comment by xurizaemon** [comment]
Gosh, "feat" feels like a lot now! Sorry 😄 but all good. We are surely going to hit a feat release with all the change that's come in anyway.

@joeyguerra just want to check you realise that -

1. 👻  `semantic-release` uses *commit messages* not *issue titles* to determine the significance of changes in a new release, and 
2. you can change existing commit messages using `git rebase` or `git commit --amend`

(I promised somewhere to show you how that was done, and haven't yet done this.)

File: src/robot.js [hunk 1]
```diff
@@ -49,7 +49,23 @@ class Robot {
...
+      info () {
+        return log.info(...arguments)
+      },
+      warning () {
+        return log.warn(...arguments)
+      },
+      error () {
+        return log.error(...arguments)
+      }
+    }
```
Comment (xurizaemon) [inline src/robot.js:66]: I'd suggest reducing this to:

```js
this.logger = log
```

Aliasing `debug` to `notice` makes it visible by default, but removes the ability to use the granularity of logging at debug level; I'm not sure that's the right thing to do. Is that a change here or to preserve historic behaviour?

Aliasing `warning` to `warn` is [built into that log package](https://github.com/medikoo/log#available-log-levels). (In fact we're pointing `warning` to the [alias `warn`](https://github.com/medikoo/log/blob/1448c93463b8ed1dbc804dcf80dbed147a50472c/lib/private/logger-prototype/index.js#L73-L76) here.)

File: src/robot.js [hunk 2]
```diff
@@ -49,7 +49,23 @@ class Robot {
...
+    // new Log(process.env.HUBOT_LOG_LEVEL || 'info')
```
Comment (xurizaemon) [inline src/robot.js:67]: This last comment line can be removed.

**Review by xurizaemon — CHANGES_REQUESTED** [review]
Requesting changes, but will move this to draft anyway based on https://github.com/hubotio/hubot/pull/1616#issuecomment-1520513193

File: src/robot.js (hunk 1 above)
Comment (joeyguerra) [reply src/robot.js:66]: I tried setting `this.logger = log` but the tests failed. Sinon complained it couldn't wrap on undefined properties. Which is why I used this technique. I'll try again to see if I missed anything.

Comment (joeyguerra) [reply src/robot.js]: I've implemented your suggestion and changed the test assertion strategy to pass.

File: src/robot.js [hunk 3]
```diff
@@ -49,7 +50,9 @@ class Robot {
       response: new Middleware(this),
       receive: new Middleware(this)
     }
-    this.logger = new Log(process.env.HUBOT_LOG_LEVEL || 'info')
+    process.env.LOG_LEVEL = process.env.LOG_LEVEL || process.env.HUBOT_LOG_LEVEL || 'notice'
```
Comment (levenleven) [inline src/robot.js:53]: Is default level change intentional (`info` -> `notice`)?

**Review by levenleven — APPROVED** [review]
LGTM

File: src/robot.js (hunk 3 above)
Comment (joeyguerra) [reply src/robot.js:53]: Yes. I saw on `log`'s [README.md](https://github.com/medikoo/log#readme)

> LOG_LEVEL
(defaults to notice) Lowest log level from which (upwards) all logs will be exposed.

And so my thinking was to just set it to the default. What do you think?

Comment (levenleven) [reply src/robot.js:53]: The `info` logs won't be printed if users don't set the env var. I guess not big deal, but probably worth mentioning in release notes

Comment (joeyguerra) [reply src/robot.js:53]: I'll change it to default to `info`.

**Review by technicalpickles — APPROVED** [review]

File: src/robot.js (hunk 1 above)
Comment (joeyguerra) [reply src/robot.js]: We can merge this as soon as you confirm the changes.

**Comment by joeyguerra** [comment]
@xurizaemon I'm trying to figure out how to resolve your review. I made the changes so it should be good.

**Review by xurizaemon — APPROVED** [review]

**Comment by xurizaemon** [comment]
Weird, sorry if that left it blocked! It looked like the relevant changes were made but maybe needed further approval after that?

**Comment by joeyguerra** [comment]
No worries. Yes, I think as a reviewer, you have to go back and click “approved” somewhere. Thank you. Gonna smerge.

**Comment by joeyguerra** [comment]
Welp. No I’m not. I had to remove node 14 from the action but this PR page thinks it’s still a check. Perhaps it’s a cache invalidation problem?

**Comment by joeyguerra** [comment]
@levenleven @technicalpickles I need help with merging this PR. I removed building with node_version: 14.x in this PR because the build is failing for node version 14 with

```
npm ci
npm WARN prepare removing existing node_modules/ before installation
npm ERR! Cannot read property 'async' of undefined
```

Since we're trying to follow the Node.js release cycle, removing building with 14 seems like the right direction. But now that check is still expected but won't pass because it's essentially been removed from the workflow file.

Any suggestions on how to move forward?

**Pull request merged**